ISPN-16329 RESP should return only RESP3 response types

jabolina opened the pull request:
https://issues.redhat.com/browse/ISPN-16329

A huge diff but it should be +OK to understand. The most important part of the PR is the classes in the `serialization` package. The diff touches all commands to replace the BiConsumer usage with the new serialization. AFAICT, there were fixes to:

- CLIENT TRACKINGINFO
- COMMAND
- HELLO
- MEMORY STATS
- HGETALL
- HINCRBY
- HINCRBYFLOAT
- SDIFF
- SINTER
- ZSCORE
- ZRANK
- ZRANGE
- ZRANDMEMBER
- ZDIFF/ZDIFFSTORE
- ZINTER/ZUNION
- CONFIG GET

Additionally, any command that includes double values in the response. I created a few issues for later, to avoid too many changes: [ISPN-16471](https://issues.redhat.com/browse/ISPN-16471), [ISPN-16460](https://issues.redhat.com/browse/ISPN-16460), and [ISPN-16474](https://issues.redhat.com/browse/ISPN-16474).

I haven't added all the types from the specification. For example, we never use "push" messages or verbatim strings. I have added big numbers, even though we don't use them. We can add the missing pieces when necessary.

To give more context. I removed every "manual" serialization<sup>1</sup>, where we write to a buffer *while* processing a request. Now we have each step in distinct places, the commands only handle a request and after the return, a place to deal with serialization. The classes for serialization are private to the `serialization` package and don't leak outside. We call the appropriate static method of the `Resp3Response` class to interact with serialization. The goal with this was just to split the concerns.

This approach has some downsides. We utilize a registry to add all the specific serializers, and during the runtime, we need to identify which one to use. To list a few of the issues:

* More operations are needed to serialize a response;
* More garbage generated while iterating in the serializers;
* Buffer resizing because we serialize the object recursively without calculating the total size.

These issues mostly affect the operations with aggregate response types (arrays, sets, hashes). Simpler operations that return primitive types should find the appropriate serializer generating less garbage and allocate the buffer with the exact size. But the issues are still there nevertheless.

Honestly, I am not attached to this specific solution. My goal was to separate the serialization from the command handling. If you have other suggestions that don't suffer from the same issues or are more lightweight, I am more than happy to apply them to this PR.

---

1. The `EXEC` and PubSub listeners still need to do some "manual" handling of objects to work. I couldn't remove all the manual pieces for these commands.

jabolina commented:
I ran a comparison with redis-benchmark with this branch and main. Running locally (so, there's noise in the system) with a single node. Warming up by running 3x `redis-benchmark -h localhost -p 11222 -n 100000 --threads 12 -d 10000 --user admin -a password -3 -t set,get` and final measurement with 1_000_000 requests:

```bash
redis-benchmark -h localhost -p 11222 -n 1000000 --threads 12 -d 10000 --user admin -a password -3 -t set,get
```

**This branch**

```text
SET:
Summary:
  throughput summary: 76846.23 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.633     0.120     0.599     0.791     1.415    16.463


GET:
Summary:
  throughput summary: 199560.95 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.192     0.016     0.167     0.383     0.727    20.911
```

**main**

```text
SET:
Summary:
  throughput summary: 78382.19 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.626     0.448     0.599     0.775     1.343    11.759
        
        
GET:
Summary:
  throughput summary: 199560.95 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.193     0.024     0.167     0.407     0.767    18.879
```


And, as a table:

| # Branch         | avg   | min   | p50   | p95   | p99   | max    |
|------------------|-------|-------|-------|-------|-------|--------|
| ISPN-16329 (GET) | 0.192 | 0.016 | 0.167 | 0.383 | 0.727 | 20.911 |
| main (GET)       | 0.193 | 0.024 | 0.167 | 0.407 | 0.767 | 18.879 |
| ISPN-16329 (SET) | 0.633 | 0.120 | 0.599 | 0.791 | 1.415 | 16.463 |
| main (SET)       | 0.626 | 0.448 | 0.599 | 0.775 | 1.343 | 11.759 |

jabolina commented:
Another big thing. I should probably create a unit test for each serializer. We didn't have a way to do that before. But I'll keep the PR open to gather suggestions.

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/commands/set/SRANDMEMBER.java:
```diff
@@ -30,6 +29,6 @@ public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
       final byte[] key = arguments.get(0);
       EmbeddedSetCache<byte[], byte[]> esc = handler.getEmbeddedSetCache();
-      return handler.stageToReturn(esc.pop(key, count, false), ctx, Consumers.COLLECTION_BULK_BICONSUMER);
+      return handler.stageToReturn(esc.pop(key, count, false), ctx);
```
no Resp3Response::set here?

rigazilla reviewed (approved):
implementation lgtm

wburns reviewed (changes requested):
I think we need to tweak this in a couple ways

1. The usage of method references, while nice looking causes a BiConsumer allocation every time it is invoked. We should cache the reference in the class directly or have a shared location for them.
2. We should also change the serializer method to add a hint for nested elements when serializing to avoid the iteration. This will also allow for using simple string serializer versus the bulk one that has to write UTF-8.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/Resp3Response.java:
```diff
@@ -0,0 +1,231 @@
+    * @param alloc Buffer pool for allocation.
+    * @see PrimitiveSerializer#SIMPLE_STRING
+    */
+   public static void simpleString(CharSequence value, ByteBufPool alloc) {
```
Explicitly, yeah. But nested objects would very likely utilize simple strings. This is tricky because I didn't find anything explicitly choosing one over the other. Some tests failed because they expected a bulk string, but it was more of a Lettuce thing than for RESP3.

jabolina commented:
Updated with the suggestions and trying to tackle the points.

1. I've created the static `BiConsumer` on the `Resp3Response` class. It is not very good-looking, but it should avoid the allocations.

The second point was a bit more tricky. I've added an explicit sub-type for serializer that deals with nested objects so they can receive a `SerializationHint` that redirects to the appropriate method. I really didn't want to open the can of deeply nested objects or homogeneous elements. This means the serialization hints only work if the structure (list, set, map) contains elements of the same type. For example, a `List<byte[]>` we can hint it is an array of bulk strings. But for lists (`List.of(byte[], Map<String, byte[]>)`) that mix types, it will fall back to iterate until finding the correct serializer.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/Resp3SerializerRegistry.java:
```diff
@@ -0,0 +1,127 @@
+    *
+    * @param serializer The custom serializer.
+    */
+   public static synchronized void addCustomSerializer(JavaObjectSerializer<?> serializer) {
```
I question if this is really worth it. At worst we could just add all these infrequently used serializers to the end of the array. I think it would be more beneficial to do this and have the `candidates` above just be a static final ResponseSerializer<?>[] instead. At worst case it just means it would take a few iterations more before we throw an exception, which shouldn't be happening anyways. But the inverse is we remove the volatile read from the AtomicReference and also can remove the Collection.

jabolina replied:
Managed to remove all this. Now, we keep only the ResponseSerializer<?>[] array statically initiated.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/LPOS.java:
```diff
@@ -53,8 +52,7 @@ public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
 
-         if (res == null || res instanceof Long) Consumers.LONG_BICONSUMER.accept((Long) res, buff);
-         else Consumers.COLLECTION_LONG_BICONSUMER.accept((Collection<Long>) res, buff);
+         Resp3Response.serialize(res, buff);
```
Looks like this command can only return a Collection<Long> or a single Long. We could add an if check and call to the appropriate serializer then. Another option is you could add a record that encapsulates the value and the serializer in the lposAndReturn method.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/RespRequestHandler.java:
```diff
@@ -128,6 +129,10 @@ public void handleChannelDisconnect(ChannelHandlerContext ctx) {
       // Do nothing by default
    }
 
+   public <E> CompletionStage<RespRequestHandler> stageToReturn(CompletionStage<E> stage, ChannelHandlerContext ctx) {
```
I question the need of this method... shouldn't all callers know the response type and can use the method with ` BiConsumer<? super E, ByteBufPool> biConsumer` argument as well?

For example MEMORY can use Resp3Response.INTEGER as its third argument.

jabolina replied:
Thinking about it now, it should be a better approach. So even when we implement we need to set the response type.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/SubscriberHandler.java:
```diff
@@ -72,18 +73,14 @@ public CompletionStage<Void> onEvent(CacheEntryEvent<Object, byte[]> entryEvent)
+            // *3 + \r\n + $7 + \r\n + message + \r\n + $ + keylength + \r\n + key + \r\n +
+            // $ + valuelength + \r\n + value + \r\n
+            int byteSize = 2 + 2 + 2 + 2 + 7 + 2 + 1 + ByteBufferUtils.stringSize(key.length)
+                  + 2 + key.length + 2 + 1 + ByteBufferUtils.stringSize(value.length) + 2 + value.length + 2;
```
Why did this need to be changed? log10 should work, is the new method supposedly faster?

jabolina replied:
TBH it was just to have that in a single method. Haven't measured the difference between the two.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/SubscriberHandler.java:
```diff
@@ -170,29 +167,20 @@ public CompletionStage<RespRequestHandler> unsubscribeAll(ChannelHandlerContext
+            counter = Math.max(0, counter + (isSubscribe ? 1 : -1));
+            List<Object> reply = List.of(type, keyChannel, counter);
+            // Mixed types in array.
+            Resp3Response.array(reply, alloc);
```
I am not sure about this, tbh. I think places that use array where it is a known number of elements that have discrete types we should instead write the array length in this method and then the elements. This way no special serializer lookup is required.

jabolina replied:
Also reworked this. Anything that uses a heterogeneous structures (map, set, list), should provide a serializer. It makes it more restrictive but should avoid cases of iterating in search of all the serializers.

wburns reviewed (changes requested):
Looking great!

wburns commented:
It appears some tests are timing out with the changes.

```
2024-09-26T18:49:12.3773350Z [TestSuiteProgress] Test failed: org.infinispan.server.resp.RespSingleNodeTest.destroy
2024-09-26T18:49:12.4789165Z Dumping threads of process 264061 to /home/runner/work/infinispan/infinispan/server/resp/threaddump-RespSingleNodeTest-20240926-1849-264061.log
2024-09-26T18:49:12.6984392Z Interrupted thread testng-RespSingleNodeTest (32).
2024-09-26T18:49:13.6234258Z Killed processes 264061
2024-09-26T18:49:13.9741939Z [INFO] 
2024-09-26T18:49:13.9742393Z [INFO] Results:
2024-09-26T18:49:13.9742981Z [INFO] 
2024-09-26T18:49:13.9743384Z [ERROR] Errors: 
2024-09-26T18:49:13.9745042Z [ERROR]   RespSingleNodeTest.testRenameWithEx:1115 » RedisCommandTimeout Command timed out after 15 second(s)
2024-09-26T18:49:13.9749689Z [ERROR]   RespSingleNodeTest.testRenamenx:1125 » RedisCommandTimeout Command timed out after 15 second(s)
2024-09-26T18:49:13.9751456Z [ERROR]   RespSingleNodeTest.testScanCount:494 » RedisCommandTimeout Command timed out after 15 second(s)
2024-09-26T18:49:13.9753178Z [ERROR]   RespSingleNodeTest.testScan:475 » RedisCommandTimeout Command timed out after 15 second(s)
```

jabolina commented:
The blocking on CI is caused by [ISPN-16699](https://issues.redhat.com/browse/ISPN-16699) (#12947). I'll rebase once that's done.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/hash/HMGET.java:
```diff
@@ -36,13 +39,19 @@ public HMGET() {
    public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
       EmbeddedMultimapPairCache<byte[], byte[], byte[]> multimap = handler.getHashMapMultimap();
       CompletionStage<Collection<byte[]>> cs = multimap.get(arguments.get(0), arguments.subList(1, arguments.size()).toArray(Util.EMPTY_BYTE_ARRAY_ARRAY))
+            .thenApply(this::wrapKeys)
```
This looks like a bug with something else?

jabolina replied:
Ah, yeah. I can create an issue/PR for it.

wburns replied:
Same PR is fine, just separate the commit so it can be tracked independently.

jabolina replied:
Added 6f5e971 for it

wburns reviewed (approved):
LGTM, just some last questions.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/SerializationHint.java:
```diff
@@ -0,0 +1,46 @@
+    * @param key The type of keys in the structure.
+    * @param value The type of values in the structure.
+    */
+   record KeyValueHint(SimpleHint key, SimpleHint value, JavaObjectSerializer<Map<?, ?>> custom) implements SerializationHint { }
```
Removing it. I had an idea for it but changed my mind and forgot to remove it.

wburns commented:
Just waiting on CI to merge.

wburns commented:
Integrated into main, thanks @jabolina !
